finality-aleph/src/network.rs

```diff
@@ -233,24 +271,51 @@ enum MetaMessage {
     AuthenticationRequest(SessionId),
 }
 
+type MessageIndex = u64;
+
+#[derive(Clone, Encode, Decode)]
+struct DataMessage<D> {
+    session_id: SessionId,
+    index: MessageIndex,
+    data: D,
+    recipient: Recipient<NodeIndex>,
+}
+
+impl<D: Encode> Hash for DataMessage<D> {
+    fn hash<H: StdHasher>(&self, state: &mut H) {
+        self.index.hash(state);
+        self.data.encode().hash(state);
+    }
+}
+
+impl<D: Encode> PartialEq for DataMessage<D> {
+    fn eq(&self, other: &DataMessage<D>) -> bool {
+        self.index == other.index && self.data.encode() == other.data.encode()
+    }
+}
+
+impl<D: Encode> Eq for DataMessage<D> {}
+
 #[derive(Clone, Encode, Decode)]
 enum InternalMessage<D: Clone + Encode + Decode> {
     Meta(MetaMessage),
-    Data(SessionId, D),
+    Data(DataMessage<D>),
 }
 
 struct SessionData<D: Clone + Encode + Decode> {
     pub(crate) data_for_user: mpsc::UnboundedSender<D>,
     pub(crate) status: SessionStatus,
     pub(crate) keychain: KeyBox,
+    messages: LruCache<DataMessage<D>, ()>,
```

Comment:
I would either make this hold hashes of DataMessage's, or make the CACHE_SIZE something like 10^4 rather than 10^6. Actually, even for hashes I would reduce the size. The smaller the cache is, the more cache friendly the cache :)

Reply:
Reducing the size.